maze_transformer/training/maze_dataset.py

```diff
@@ -114,265 +125,272 @@ def n_tokens(self) -> int:
+        if fmt == SaveFormats.OBJECTS:
+            return self.mazes[index]
+        elif fmt == SaveFormats.TOKENS:
+            return maze_to_tokens(self.mazes[index], self.cfg.node_token_map)
+        elif fmt == SaveFormats.ARRAY:
+            raise NotImplementedError("getting as array not implemented yet")
+        else:
             raise ValueError(
-                "at least one of mazes_objs, mazes_tokens, mazes_tokenized must be provided to MazeDataset"
+                f"unknown fmt {fmt}, expected an instance of `SaveFormats` enum"
             )
 
-        # transfer
-        self.mazes_objs: list[SolvedMaze] | None = mazes_objs
-        self.mazes_tokens: list[list[str]] | None = mazes_tokens
-        self.mazes_array: IndexedArray | None = mazes_array
-
-        # process into tokens
-        if (self.mazes_objs is not None) and (self.mazes_tokens is None):
-            with multiprocessing.Pool() as pool:
-                self.mazes_tokens = list(
-                    tqdm(
-                        pool.map(
-                            partial(maze_to_tokens, node_token_map=cfg.node_token_map),
-                            self.mazes_objs,
-                        ),
-                        total=len(self.mazes_objs),
-                        desc="tokenizing mazes",
-                        unit="maze",
-                    )
-                )
-
-        # process tokens into tokenized
-        if (self.mazes_tokens is not None) and (mazes_array is None):
-            max_len: int = max(len(t) for t in self.mazes_tokens)
-            if max_len > cfg.seq_len_max:
-                raise ValueError(f"{max_len=} exceeds {cfg.seq_len_max=}")
-
-            self.mazes_array = IndexedArray.from_sequences(
-                [cfg.tokenize_seq(m) for m in self.mazes_tokens]
-            )
-
-        # validate
-        # if any(x is None for x in (self.mazes_objs, self.mazes_tokens, self.mazes_tokenized)):
-        # 	raise ValueError(f"MazeDataset invalid, something is None: {type(self.mazes_objs) = } {type(self.mazes_tokens) = } {type(self.mazes_tokenized) = }")
-
-        if self.mazes_objs is not None and self.mazes_tokens is not None:
-            if len(self.mazes_objs) != len(self.mazes_tokens):
-                raise ValueError(
-                    f"MazeDataset invalid: {len(self.mazes_objs) = }, {len(self.mazes_tokens) = }"
-                )
+    def __getitem__(self, index: int, pad: bool = True) -> str:
+        """index into mazes_array.arr, getting from the start of the correct sequence, padding if necessary"""
 
-        if self.mazes_objs is not None and self.mazes_array.indices is not None:
-            if len(self.mazes_objs) != len(self.mazes_array.indices):
-                raise ValueError(
-                    f"MazeDataset invalid: {len(self.mazes_objs) = }, {len(self.mazes_array.indices) = }"
-                )
+        return " ".join(maze_to_tokens(self.mazes[index], self.cfg.node_token_map))
```

Comment:
This is fine, but it means we will need to provide a collation function for the pytorch dataloader to do the batching (as we aren't padding inside of the dataset __getiem__. This is _probably_ the most efficient way to do it tbf. 

That being said, it is fairly amusing that we start from a file of maze strings which we convert to `SolvedMaze` objects, then reconvert those into tokens with `maze_to_tokens` which in turn gets joined into a full string, then collated, and then batched by the `wrapped_tokenizer`.

Reply:
After playing around with this a bit I'm leaning towards returning a SolvedMaze here, since that's the most generally useful thing - we can have lots of methods on it to easily convert to other formats. Then as you say, do the conversion to tokens in the training loop - this can be more efficient because we can operate on the whole batch.

Reply:
Indeed It would be good to have a standard representation for mazes which we pass around, and then all the conversion happens inside plotting functions or the tokenizer.

Reply:
Yep good point - I need to think about how to handle that in a nice way.

Reply:
`SolvedMaze` now inherits from `LatticeMaze`, but is assumed to actually have a `solution` component. Maybe then we should return a `LatticeMaze`, and perform a check downstream to ensure the object is actually an instance of `SolvedMaze`?

Question for @afspies : do we want to assume that the mazes come with a start and end point? If so, its trivial to then compute a solution (assuming one is possible).

Reply:
Hmm yeah if we need to support mazes without solutions then it may indeed be better to roll the `TargettedMaze` and `SolvedMaze` back into `LatticeMaze`. Having them seperate has been causing us headaches anyway - needing to do `if maze.solution` in a bunch of places may be the lesser evil here.

Btw @mivanit I’ve been working on some of these discussed changes [here](https://github.com/AISC-understanding-search/maze-transformer/tree/zanj-datasets-getitem) - haven’t wanted to put the commits in this PR yet because they’re still very WIP, just letting you know so we don’t duplicate effort.